test/replication-luatest/election_fencing_test.lua

```diff
@@ -0,0 +1,288 @@
+    -- replicate previously "frozen" synchronous transactions and confirm them.
+    box_cfg_update({g.server_2}, {replication = g.box_cfg.replication})
+    wait_connected(g.server_1, g.server_2)
+    g.server_1:wait_election_leader()
+    wait_sync(g.server_1, {g.server_2})
+    luatest.helpers.retrying({}, function()
+        luatest.assert(g.server_1:exec(function()
+            return box.info.synchro.queue.len == 0
+        end))
+    end)
+
+    local ret = g.server_1:exec(function()
+        return box.space.test:select{1}
+    end)
+    luatest.assert_equals(ret, {{1}},
+        'Sync write confirmed after leadership regain')
+
+    box_cfg_update({g.server_1}, {
+        election_fencing_enabled = false,
+        replication_synchro_timeout = LONG_TIMEOUT,
+    })
+
+    box_cfg_update({g.server_1}, {replication = {}})
+    box_cfg_update({g.server_2, g.server_3}, {replication = {
+        g.server_2.net_box_uri,
+        g.server_3.net_box_uri,
+    }})
+    luatest.helpers.retrying({}, function()
+        luatest.assert(g.server_1:exec(function()
+            return (box.info.replication[2].upstream == nil and
+                    box.info.replication[3].upstream == nil and
+                    box.info.replication[2].downstream.status == 'stopped' and
+                    box.info.replication[3].downstream.status == 'stopped')
+        end))
+    end)
+    wait_connected(g.server_2, g.server_3)
+
+    local leader, limbo_len = g.server_1:exec(function(t)
+        require('fiber').sleep(t)
+        return box.info.election.leader, box.info.synchro.queue.len
+    end, {DEATH_TIMEOUT})
+    luatest.assert_equals(leader, 1)
+    luatest.assert_equals(limbo_len, 0)
+
+    g.server_1:exec(function()
+        require('fiber').create(function() box.space.test:replace{2} end)
+    end)
+
+    box_cfg_update({g.server_1}, {election_fencing_enabled = true})
+    box_cfg_update({g.server_1}, {replication_synchro_timeout = SHORT_TIMEOUT})
+
+    local leader, limbo_len = g.server_1:exec(function(t)
+        require('fiber').sleep(t)
+        return box.info.election.leader, box.info.synchro.queue.len
+    end, {DEATH_TIMEOUT})
+    luatest.assert_equals(leader, 0)
+    luatest.assert_equals(limbo_len, 1)
+    box_cfg_update({g.server_2}, {election_mode = 'candidate'})
+
+    promote(g.server_2)
+    box_cfg_update(g.cluster.servers, {replication = g.box_cfg.replication})
+
+    -- If a new leader was elected while old one was fenced - "frozen"
+    -- transactions should be rollbacked by new leader.
+    luatest.helpers.retrying({}, function()
+       luatest.assert(g.server_1:exec(function()
+            return box.info.synchro.queue.len == 0
+        end))
+    end)
+
+    wait_sync(g.server_2, {g.server_1})
+    local ret = g.server_1:exec(function()
+        return box.space.test:select{2}
+    end)
+    luatest.assert_equals(ret, {},
+        'Sync write rollbacked after new leader discovered')
+end
```

Comment:
Imagine the following scenario: there were 3 instances, huge, each is 500 GB of in-memory data. Restart of any of them takes hours or even days. The fencing was enabled. Now the leader starts a transaction and suddenly the other 2 nodes are lost completely, destroyed or corrupted to a non-recoverable state, or were split-brained and have conflicting data which we want to discard. Anyway, we want this leader to proceed alone. And we don't want to restart it. It must remain available for reading at least.

For that we want to forcefully reduce quorum to 1 temporarily, unfreeze the limbo, and continue writing while other nodes rejoin.

Here I tried to repro that scenario (but with just 2 nodes for simplicity):
**Instance 1**
```Lua
--
-- Step 1
--
fiber = require('fiber')
box.cfg{
    listen = 3313,
    replication = {3313, 3314},
    replication_synchro_timeout = 1000,
    replication_synchro_quorum = 2,
    election_mode = 'candidate',
}
box.schema.user.grant('guest', 'super')
s = box.schema.create_space('test', {is_sync = true})
_ = s:create_index('pk')

--
-- Step 3
-- Start writing something.
--
box.error.injection.set('ERRINJ_WAL_DELAY', true)
f = fiber.create(s.replace, s, {1})

--
-- Step 5
--
-- Firstly, I reduce the timeout to see the txn outcome sooner.
box.cfg{replication_synchro_timeout = 1}
-- Then I disable the fencing as an emergency.
box.cfg{election_fencing_enabled = false}
-- I want to become a rightful leader again and confirm txns alone.
box.cfg{replication_synchro_quorum = 1}
```

**Instance 2**
```Lua
--
-- Step 2
--
box.cfg{
    listen = 3314,
    replication = {3313, 3314},
    election_mode = 'voter',
}

--
-- Step 4
-- The node suddenly dies, leaving the leader with a txn in its limbo.
--
os.exit(0) -- Or just Ctrl+D.
```

But after step 4 nothing happens:
```
tarantool> 2022-04-23 00:42:37.751 [66514] main I> RAFT: begin new election round
2022-04-23 00:42:37.751 [66514] main I> RAFT: bump term to 3, follow
2022-04-23 00:42:37.751 [66514] main I> RAFT: vote for 1, follow

---
...

tarantool> f
---
- status: suspended
  name: lua
  id: 123
...
```
The transaction still hangs. I tried to call `box.ctl.promote()` but it also hung. Why? If this is really supposed to happen, I would advise against such behaviour - being able to restore an instance to a functional state without a restart is important during emergencies like in the case I described above.

Reply:
Did you forget to disable ERRINJ_WAL_DELAY in step 5? If I disable it I get something like this:
```
tarantool> box.error.injection.set('ERRINJ_WAL_DELAY', false)
---
- ok
...

tarantool> 2022-04-23 08:51:34.837 [2368154] main txn.c:842 W> too long WAL write: 1 rows at LSN 10: 13.527 sec
2022-04-23 08:51:34.838 [2368154] relay/127.0.0.1:37406/101/main coio.c:348 E> SocketError: unexpected EOF when reading from socket, called on fd 19, aka 127.0.0.1:3301, peer of 127.0.0.1:37406: Broken pipe
2022-04-23 08:51:34.838 [2368154] relay/127.0.0.1:37406/101/main I> exiting the relay loop

---
...

tarantool> f
---
- status: suspended
  name: lua
  id: 124
...

tarantool> box.cfg{election_fencing_enabled = false}
2022-04-23 08:51:59.141 [2368154] main/103/1.lua I> RAFT: fencing disabled
2022-04-23 08:51:59.141 [2368154] main/103/1.lua I> set 'election_fencing_enabled' configuration option to false
---
...

tarantool> f
---
- status: suspended
  name: lua
  id: 124
...

tarantool> box.cfg{replication_synchro_quorum = 1}
2022-04-23 08:52:12.651 [2368154] main/103/1.lua I> set 'replication_synchro_quorum' configuration option to 1
---
...

tarantool> 2022-04-23 08:52:12.651 [2368154] main/124/lua txn_limbo.c:288 E> ER_SYNC_QUORUM_TIMEOUT: Quorum collection for a synchronous transaction is timed out

---
...

tarantool> f
---
- the fiber is dead
...

tarantool> 2022-04-23 08:52:16.653 [2368154] main I> RAFT: begin new election round
2022-04-23 08:52:16.653 [2368154] main I> RAFT: bump term to 3, follow
2022-04-23 08:52:16.653 [2368154] main I> RAFT: vote for 1, follow
2022-04-23 08:52:16.653 [2368154] main/113/raft_worker I> RAFT: persisted state {term: 3, vote: 1}
2022-04-23 08:52:16.653 [2368154] main/113/raft_worker I> RAFT: enter leader state with quorum 1

---
...
```

Reply:
UPD. If I want to confirm hanging transactions I can decrease quorum without turning fencing off. Node will become leader and confirm all waiting transactions.
```
tarantool> box.error.injection.set('ERRINJ_WAL_DELAY', false)
---
- ok
...

tarantool> 2022-04-23 08:51:34.837 [2368154] main txn.c:842 W> too long WAL write: 1 rows at LSN 10: 13.527 sec
2022-04-23 08:51:34.838 [2368154] relay/127.0.0.1:37406/101/main coio.c:348 E> SocketError: unexpected EOF when reading from socket, called on fd 19, aka 127.0.0.1:3301, peer of 127.0.0.1:37406: Broken pipe
2022-04-23 08:51:34.838 [2368154] relay/127.0.0.1:37406/101/main I> exiting the relay loop

---
...

tarantool> f
---
- status: suspended
  name: lua
  id: 124
...

tarantool> box.cfg{replication_synchro_quorum = 1}
2022-04-23 09:05:59.124 [2374667] main/103/1.lua I> set 'replication_synchro_quorum' configuration option to 1
---
...

tarantool> 2022-04-23 09:06:03.132 [2374667] main I> RAFT: begin new election round
2022-04-23 09:06:03.132 [2374667] main I> RAFT: bump term to 3, follow
2022-04-23 09:06:03.132 [2374667] main I> RAFT: vote for 1, follow
2022-04-23 09:06:03.132 [2374667] main/113/raft_worker I> RAFT: persisted state {term: 3, vote: 1}
2022-04-23 09:06:03.132 [2374667] main/113/raft_worker I> RAFT: enter leader state with quorum 1

---
...

tarantool> f
---
- the fiber is dead
...

tarantool> box.space.test:select()
---
- - [1]
...
```

Reply:
Indeed, I a stupid mistake, I forgot to enable WAL again. Then it looks good, thanks for checking this!